src/transformers/feature_extraction_utils.py

```diff
@@ -175,27 +176,56 @@ def as_tensor(value):
         return self
 
     @torch_required
-    # Copied from transformers.tokenization_utils_base.BatchEncoding.to with BatchEncoding->BatchFeature
-    def to(self, device: Union[str, "torch.device"]) -> "BatchFeature":
+    def to(self, *args, **kwargs) -> "BatchFeature":
         """
-        Send all values to device by calling `v.to(device)` (PyTorch only).
+        Send all values to device by calling `v.to(*args, **kwargs)` (PyTorch only). This should support casting in
+        different `dtypes` and sending the `BatchFeature` to a different `device`.
 
         Args:
-            device (`str` or `torch.device`): The device to put the tensors on.
+            args (`Tuple`):
+                Will be passed to the `to(...)` function of the tensors.
+            kwargs (`Dict`, *optional*):
+                Will be passed to the `to(...)` function of the tensors.
 
         Returns:
             [`BatchFeature`]: The same instance after modification.
         """
-
-        # This check catches things like APEX blindly calling "to" on all inputs to a module
-        # Otherwise it passes the casts down and casts the LongTensor containing the token idxs
-        # into a HalfTensor
-        if isinstance(device, str) or is_torch_device(device) or isinstance(device, int):
-            self.data = {k: v.to(device=device) for k, v in self.data.items()}
-        else:
-            logger.warning(f"Attempting to cast a BatchFeature to type {str(device)}. This is not supported.")
+        import torch  # noqa
+
+        new_data = {}
+        device = self._parse_to_args(*args, **kwargs)
+        # We cast only floating point tensors to avoid issues with tokenizers casting `LongTensor` to `FloatTensor`
+        for k, v in self.items():
+            # check if v is a floating point
+            if torch.is_floating_point(v):
+                # cast and send to device
+                new_data[k] = v.to(*args, **kwargs)
+            else:
+                # Just send to device for int tensors
+                if device is not None:
+                    new_data[k] = v.to(device=device)
+                else:
+                    new_data[k] = v
+        self.data = new_data
         return self
 
+    def _parse_to_args(self, *args, **kwargs):
```

Comment:
Nice catch! It's because on [af04179](https://github.com/huggingface/transformers/pull/20536/commits/af0417936b8472f2a2114dc61e1fec06573147a2) I used to extract the `dtype` too until I figured out that the `dtype` was not needed at all :D